lib/semantics/check-omp-clause.cc

```diff
@@ -0,0 +1,164 @@
+//                        num-threads-clause |
+//                        default-clause |
+//                        private-clause |
+//                        firstprivate-clause |
+//                        shared-clause |
+//                        copyin-clause |
+//                        reduction-clause |
+//                        proc-bind-clause
+void OmpClauseChecker::Enter(const parser::OmpBlockDirective::Parallel &x) {
+  std::set<std::type_index> s{typeid(parser::OmpIfClause),
+      typeid(parser::OmpClause::NumThreads), typeid(parser::OmpDefaultClause),
+      typeid(parser::OmpClause::Private),
+      typeid(parser::OmpClause::Firstprivate),
+      typeid(parser::OmpClause::Shared), typeid(parser::OmpClause::Copyin),
+      typeid(parser::OmpReductionClause), typeid(parser::OmpProcBindClause)};
+  SetAllowed(s);
+}
+
+void OmpClauseChecker::Enter(const parser::OmpBlockDirective &x) {
+  SetCurrentDirectiveSource(x.source);
+}
+
+void OmpClauseChecker::Leave(const parser::OmpEndBlockDirective &x) {
+  EmptyAllowed();
+  EmptyCurrentDirectiveSource();
+  EmptyCurrentClauseSource();
+}
+
+void OmpClauseChecker::Enter(const parser::OmpClause &x) {
+  SetCurrentClauseSource(x.source);
+}
+
+void OmpClauseChecker::Enter(const parser::OmpClause::Defaultmap &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Inbranch &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Mergeable &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Nogroup &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Notinbranch &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Untied &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Collapse &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Copyin &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Copyprivate &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Device &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::DistSchedule &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Final &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Firstprivate &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::From &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Grainsize &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::Lastprivate &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::NumTasks &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::NumTeams &x) {
+  CheckAllowed(x);
+}
+void OmpClauseChecker::Enter(const parser::OmpClause::NumThreads &x) {
```

Comment:
This is a good question. I will implement the repetition check with `std::bitset`. To make the framework more complete, the stack needs to be introduced for OpenMP constructs and `type_index` will be replaced with `EnumSet`. So, this pull request is on hold for now.

Reply:
There are at least two possibilities for implementing a stack.

One is to put an explicit stack container class of directives in the semantics context_, and do pushes and pops in directive Enter and Leave functions.

Another is to maintain a current directive reference in the semantics context_, along with a parent link in a directive subcontext class.  When a nested directive is encountered in an Enter function, it stores the current directive in its own context (push), and reinstates (pops) it when Leave-ing.  That way, all clause Enter/Leave functions automatically apply to the current directive, and the stack per se is only implicit.

In either case, to implement (4.5 page 227) region nesting restrictions, if directive subcontext contains a summary of the types of all enclosing directives, it may never be necessary to actually search down in an explicit stack or chase parent directive links to validate nesting restrictions.  The pgi compiler does something like this with the DI_NEST macro for OpenMP/loop/if/... scopes in comp/fe90/shared/src/semant.h.

Reply:
Thanks Val, please check the latest `OmpContext` implementation.

Reply:
It should be ok to use an explicit stack.